Include shlwapi.h explicitly on Windows

mkitti opened the pull request:
This fixes builds on mingw32 and other compilers.

I suspect this will fix https://github.com/HDFGroup/hdf5/issues/2396 , #2408

mkitti committed:
Include shlwapi.h explicitly

mkitti commented:
xref: https://learn.microsoft.com/en-us/windows/win32/api/shlwapi/nf-shlwapi-strstria

mkitti committed:
Add shlwapi library to LINK_LIB when header is detected

Looking for StrStrIA in the shlwapi library is not reliable due
to stdcall on mingw32.

mkitti commented:
<strike>Hmm, this alone may not be sufficient. `shlwapi` may also needed to be added to `LINK_LIB` when the header is detected. Testing again.</strike>

edit: [d300feb](https://github.com/HDFGroup/hdf5/pull/2407/commits/d300febacda44b5f81d7d882c6f45b56c0424c97) is sufficient, I just forgot to apply the patch.

mkitti commented:
During configuration this still occurs in the mingw32 environment.

```
-- Looking for StrStrIA in shlwapi;m;ws2_32;wsock32
-- Looking for StrStrIA in shlwapi;m;ws2_32;wsock32 - not found
```

This is due to name mangling that occurs in 32-bit windows with stdcall:
```
$ nm /mingw64/lib/libshlwapi.a  | grep StrStrIA
0000000000000000 I __imp_StrStrIA
0000000000000000 T StrStrIA

$ nm /mingw32/lib/libshlwapi.a  | grep StrStrIA
00000000 I __imp__StrStrIA@8
00000000 T _StrStrIA@8
```

hmaarrfk commented:
Do you think it cleanly applies to 1.14.0? 
If so I'm going to test it in https://github.com/conda-forge/hdf5-feedstock/pull/188 if that is ok.

hmaarrfk commented:
I can confirm that this fixes things for the "windows" compiler, at least in terms of the segfault.

hmaarrfk commented:
didn't you want to fix the `mingw32` configuration?

mkitti commented:
> didn't you want to fix the `mingw32` configuration?

This addresses the issues for Windows-based compilation in general actually. Was there something else that needed to be addressed here?

hmaarrfk commented:
No. I'm just glad I got to test the fix before you merged anything.

emmenlau commented:
I think this PR opened a problem, because `shlwapi` is now a transitive dependency of hdf5, but this is not exposed via the cmake properties. Linking hdf5 on Windows leads to errors that can be resolved by linking `shlwapi`. But in cmake world, this dependency should be exposed from the cmake target.

So, since this commit, the following is no longer sufficient to link hdf5 for me:
```
  find_package(HDF5 MODULE REQUIRED COMPONENTS C)

  target_include_directories(libdeps SYSTEM INTERFACE ${HDF5_INCLUDE_DIRS})
  target_link_libraries(libdeps INTERFACE ${HDF5_C_LIBRARIES}) 
  target_compile_definitions(libdeps INTERFACE ${HDF5_DEFINITIONS})
```

mkitti commented:
@emmenlau I think the problem would be easier to track if to created a new issue.
